### DISCUSSION TITLE
How to load data from JSON?

### DISCUSSION THREAD

**Discussion by andyjessop** [toeverything/blocksuite#5330] [General]
You can get the JSON from, e.g. `$0.workspace.doc.toJSON()`. Let's say you then save that in a DB and want to load it later. What is the command to load that back into the `doc`? I can't see this in the docs anywhere.

**Comment by doodlewind** [comment]
Thanks for the feedback! The `workspace.doc.toJSON()` is simply `ydoc.toJSON()`, which should only be used for debug, since it expresses all rich text content as plain text (see [upstream discussion](https://github.com/yjs/yjs/issues/284#issuecomment-1406620927)). The best practice for JSON serializing is to use following:

``` ts
import { Job, Workspace } from '@blocksuite/store';

const workspace = new Workspace();
const job = new Job({ workspace });

const json = await job.pageToSnapshot(page);

const newPage = await job.snapshotToPage(json);
```

You can also use the `job` variable in playground console, and use the `Import Snapshot` / `Export Snapshot` entries here for testing zip snapshot bundle:

<img width="735" alt="image" src="https://github.com/toeverything/blocksuite/assets/7312949/c480bc61-a790-466f-9df8-edf61ba0dba5">

But snapshot leads to page-level importing and exporting. BlockSuite also supports provider-based incremental state update over WebSocket and IndexedDB. In this case, you can simply store the CRDT binary (somewhat like encoded protobuf).

I plan to rewrite the https://github.com/toeverything/blocksuite/blob/master/packages/docs/data-persistence.md doc this week to clarify the data persistence matters.
